File: commons/javafx/src/main/java/com/cedarsoft/commons/javafx/NumberStringConverterForLongHex.java
package com.cedarsoft.commons.javafx;

import java.text.NumberFormat;

import javax.annotation.Nonnull;

import javafx.util.converter.NumberStringConverter;

/**
 * A {@link NumberStringConverter} for long values only, in hex representation
 */
public class NumberStringConverterForLongHex extends NumberStringConverter {

  @Nonnull
  private final NumberFormat integerHexNumberFormat; // works for longs, too

  public NumberStringConverterForLongHex() {
    integerHexNumberFormat = new HexNumberFormat();
  }

  @Nonnull
  @Override
  protected NumberFormat getNumberFormat() {
    return integerHexNumberFormat;
  }

  @Override
  public Number fromString(String value) {
    if ("-".equals(value)) {
      return 0;
    }

    if (value == null) {
      return null;
    }

    if (value.startsWith("0x")) {
      return Long.parseLong(value.substring(2), 16);
    }
    return Long.parseLong(value, 16);
  }
}

File: commons/javafx/src/main/kotlin/com/cedarsoft/commons/javafx/NumberStringConverterForLongHex.kt
package com.cedarsoft.commons.javafx

import javafx.util.converter.NumberStringConverter
import java.text.NumberFormat

/**
 * A [NumberStringConverter] for long values only, in hex representation
 */
class NumberStringConverterForLongHex : NumberStringConverter() {

  // works for longs, too
  private val integerHexNumberFormat: NumberFormat = HexNumberFormat()

  override fun getNumberFormat(): NumberFormat {
    return integerHexNumberFormat
  }

  override fun fromString(value: String?): Number? {
    if ("-" == value) {
      return 0
    }
    if (value == null) {
      return null
    }
    return if (value.startsWith("0x")) {
      value.substring(2).toLong(16)
    } else value.toLong(16)
  }
}
